src/randomness/RandomnessSender.sol

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+import {AccessControlEnumerableUpgradeable} from
+    "@openzeppelin/contracts-upgradeable/access/extensions/AccessControlEnumerableUpgradeable.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {UUPSUpgradeable} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+import {TypesLib} from "../libraries/TypesLib.sol";
+
+import {IRandomnessReceiver} from "../interfaces/IRandomnessReceiver.sol";
+import {IRandomnessSender} from "../interfaces/IRandomnessSender.sol";
+import {ISignatureSender} from "../interfaces/ISignatureSender.sol";
+
+import {SignatureReceiverBase} from "../signature-requests/SignatureReceiverBase.sol";
+
+contract RandomnessSender is
+    IRandomnessSender,
+    SignatureReceiverBase,
+    Initializable,
+    UUPSUpgradeable,
+    AccessControlEnumerableUpgradeable
+{
+    // the DST is used to separate randomness being used as signatures for other things
+    string public constant DST = "randomness:0.0.1:bn254";
+    string public constant SCHEME_ID = "BN254";
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+
+    uint256 public nonce = 0;
+
+    // Mapping from randomness requestID to callbacks
+    mapping(uint256 => TypesLib.RandomnessRequest) private callbacks;
+    // array of all requests
+    TypesLib.RandomnessRequest[] private allRequests;
```

Comment:
is it worth following the same pattern as we do for timelock requests? seems like having an `error` array is a good way to ensure reliability

Reply:
ah wait, I see that it's managed in the `SignatureSender` - that makes more sense actually